k2/csrc/fsa.h

```diff
@@ -77,13 +77,40 @@ struct Fsa {
   // contains the first arc-index leaving this state (index into `arcs`).
   // The next element of this array gives the end of that range.
   // Note: the final-state is numbered last, and implicitly has no
-  // arcs leaving it.
+  // arcs leaving it. For non-empty FSA, we put a duplicate of the final state
+  // at the end of `arc_indexes` to avoid boundary check for some FSA
+  // operations. Caution: users should never call `arc_indexes.size()` to get
+  // the number of states, they should call `NumStates()` to get the number.
   std::vector<int32_t> arc_indexes;
 
   // Note: an index into the `arcs` array is called an arc-index.
   std::vector<Arc> arcs;
 
-  StateId NumStates() const { return static_cast<StateId>(arc_indexes.size()); }
+  Fsa() {}
+  // just for creating testing FSA examples for now.
+  Fsa(std::vector<Arc> fsa_arcs, int32_t final_state) {
```

Comment:
`final_state` is not needed. It can be inferred from the `arcs` array.

Reply:
We can't.  Just imagine an example with all states are not co-accessible.

Reply:
Shouldn't the final state **always** have the largest state number?

Reply:
Maybe, but just image an FSA with arc `0->1` and final state `3`. 
We can't get how many states from `arcs`.